File: AppEngine/src/test/java/net/twisterrob/blt/gapp/viewmodel/GwenStatus.java
package net.twisterrob.blt.gapp.viewmodel;

import java.util.*;

import static org.mockito.Mockito.*;

import com.shazam.gwen.collaborators.Arranger;

import net.twisterrob.blt.io.feeds.trackernet.LineStatusFeed;
import net.twisterrob.blt.io.feeds.trackernet.model.*;
import net.twisterrob.blt.model.Line;

class GwenStatus implements Arranger {
	private final LineStatusFeed feed = mock(LineStatusFeed.class);
	private final Map<Line, LineStatus> statuses = new EnumMap<>(Line.class);

	public GwenStatus contains(Line line, String description) {
		LineStatus status = new LineStatus();
		status.setLine(line);
		status.setType(DelayType.Unknown);
		status.setDescription(description);
		statuses.put(line, status);
		return this;
	}

	public GwenStatus contains(Line line, DelayType disruption) {
		LineStatus status = new LineStatus();
		status.setLine(line);
		status.setType(disruption);
		statuses.put(line, status);
		return this;
	}

	public GwenStatus doesNotContain(Line line) {
		statuses.remove(line);
		return this;
	}

	public Result createResult() {
		when(feed.getStatusMap()).thenReturn(statuses);
		return new Result(new Date(), feed);
	}
}

File: web/status-history/src/test/java/net/twisterrob/travel/statushistory/viewmodel/GwenStatus.kt
package net.twisterrob.travel.statushistory.viewmodel

import com.shazam.gwen.collaborators.Arranger
import net.twisterrob.blt.model.DelayType
import net.twisterrob.blt.model.Line
import net.twisterrob.blt.model.LineStatus
import net.twisterrob.blt.model.LineStatus.BranchStatus
import net.twisterrob.blt.model.LineStatuses
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import java.util.Date
import java.util.EnumMap

internal class GwenStatus : Arranger {

	private val feed: LineStatuses = mock()
	private val statuses: MutableMap<Line, LineStatus> = EnumMap(Line::class.java)

	fun contains(line: Line, description: String?): GwenStatus = apply {
		statuses[line] = LineStatus(
			line = line,
			type = DelayType.Unknown,
			isActive = true,
			branchStatuses = emptyList(),
			description = description,
		)
	}

	fun contains(line: Line, description: String?, vararg branches: BranchStatus): GwenStatus = apply {
		statuses[line] = LineStatus(
			line = line,
			type = DelayType.Unknown,
			isActive = true,
			description = description,
			branchStatuses = branches.toList(),
		)
	}

	fun contains(line: Line, disruption: DelayType): GwenStatus = apply {
		statuses[line] = LineStatus(
			line = line,
			type = disruption,
			isActive = true,
			description = null,
			branchStatuses = emptyList(),
		)
	}

	fun doesNotContain(line: Line): GwenStatus = apply {
		statuses.remove(line)
	}

	fun createResult(): Result {
		`when`(feed.statuses).thenReturn(statuses.values.toList())
		return Result.ContentResult(Date(), feed)
	}
}
